kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,137 @@
+    rule [mask-b-and]:
+      MASK:Int &Int SLOT:Int =>
+        #asWord ( #buf ( 32, SLOT ) [ 32 -Int ( #getMaskShiftBytes(MASK) +Int #getMaskWidthBytes(MASK) ) := #buf ( #getMaskWidthBytes(MASK), 0 ) ] )
+        requires #isMask(MASK) andBool #rangeUInt(256, SLOT)
+        [simplification, concrete(MASK), preserves-definedness]
+
+    // 2a. |Int and +Bytes, update to be done in left
+    rule [bor-update-to-left]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( #buf ( 32 -Int lengthBytes(B2), (A /Int (2 ^Int (8 *Int lengthBytes(B2)))) |Int #asWord ( B1 ) ) +Bytes B2 )
+        requires #rangeUInt(256, A) andBool A modInt (2 ^Int (8 *Int lengthBytes(B2))) ==Int 0 andBool lengthBytes(B1 +Bytes B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 2b. |Int of +Bytes, update to be done in right
+    rule [bor-update-to-right]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( B1 +Bytes #buf ( lengthBytes(B2), A |Int #asWord ( B2 ) ) )
+        requires 0 <=Int A andBool A <Int 2 ^Int (8 *Int lengthBytes(B2)) andBool lengthBytes(B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 3a. Update with explicit shift and symbolic slot
+    rule [bor-update-with-shift]:
+      ( SHIFT *Int X ) |Int Y => #asWord ( #buf( 32 -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, Y ) )
+      requires #isByteShift(SHIFT)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (32 -Int ( log2Int(SHIFT) /Int 8 )))
+       andBool 0 <=Int Y andBool Y <Int SHIFT
+       [simplification, concrete(SHIFT), comm, preserves-definedness]
+
+    // 3b. Buffer cropping
+    rule [buf-asWord-crop]:
+      #buf (W:Int , #asWord(B:Bytes)) => #range(B, lengthBytes(B) -Int W, W)
+      requires 0 <=Int W andBool W <=Int 32 andBool W <Int lengthBytes(B)
+       andBool #asWord ( #range(B, 0, lengthBytes(B) -Int W) ) ==Int 0
+      [simplification, concrete(W), preserves-definedness]
+
+    // 3c. Splitting the updated buffer into the updated value and the trailing zeros, explicit shift
+    rule [buf-split-l]:
+      #buf ( W, SHIFT *Int X ) => #buf( W -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, 0)
+      requires 0 <=Int W andBool W <=Int 32 andBool #isByteShift(SHIFT)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W -Int ( log2Int(SHIFT) /Int 8)))
+       [simplification, concrete(W, SHIFT), preserves-definedness]
+
+    // 3d. Splitting the updated buffer into the updated value and the trailing zeros, implicit shift
+    rule [bor-split]:
+      X |Int Y => #asWord ( #buf ( 32 -Int #getFirstOneBit(X) /Int 8, X /Int ( 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) ) ) ) +Bytes
+                            #buf ( #getFirstOneBit(X) /Int 8, Y ) )
+      requires #rangeUInt(256, X) andBool 0 <=Int #getFirstOneBit(X)
+       andBool 0 <=Int Y andBool Y <Int 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) )
+       [simplification, concrete(X), preserves-definedness]
```

Comment:
It seems that this generalizes 3a completely (`bor-update-with-shift`). Is there any reason to keep 3a then? Or, if 3a and this both apply, maybe we prefer the result we get with 3a, in which case we make this rule lower priority `simplification(60)`?

Reply:
Although, this says `concrete(X)`, which is not more general.... so maybe both are needed.

Reply:
This one is sort of a last resort, when `VALUE` is concrete and so `SHIFT *Int VALUE` is concrete and so we can't see the shift immediately. I think both are needed.

Reply:
Maybe then `simplification(60)` so that the other lemmas are prioritized over it if they also apply?